### PR TITLE
Addition of depthwise Separable Convolution layer

### PR DESCRIPTION
**Pull request by SIDHARTH-7**
this PR is to restart an old PR: #3007 which was being worked on by @Aakash-kaushik , @kartikdutt18 and @zoq 
I have added some files from the above PR

@shrit please review the PR

### CODE REVIEW

**Comment by shrit** [comment]
@SIDHARTH-7, this starts to look good, please add a couple of tests so we can review it.

**Comment by SIDHARTH-7** [comment]
@shrit I was checking the tests and there are some missing values in tests for the `Backward()` function and I checked the tests of some other layers in mlpack which just check dimensions of the `input`,`output` and `grad` will it be better if I add tests similar to those rather than comparing the sum of the `output `matrices.
Screenshot 2023-12-15 212631
do check the last line in this image

**Comment by SIDHARTH-7** [comment]
@shrit any idea why these checks are failing

**Comment by shrit** [comment]
@SIDHARTH-7 please have a look at the errors, they are fairly simple to understand, you need to adapt the Separable Conv layer code to make it look like mlpack 4.0, we have done a lot of refactoring between mlpack 3.4 and mlpack 4. The code you have is basically copied from mlpack 3.4 and will not run directly on mlpack 4. We have changed namespaces, introduced inheritance, removed `boost::variants`, and many other things that I can not recall.

The Separable class need some refactoring, we would have merged it in the past if it was ready.
please do what @rcurtin mentioned in the chat regarding refactoring and also add this link:
https://github.com/mlpack/mlpack/pull/3547
Once the refactoring is done this should pass by itself

**Comment by SIDHARTH-7** [comment]
@shrit I have Finished the refactoring and it passed all checks , Please tell me if anything else needs to be done.

File: src/mlpack/methods/ann/layer/separable_convolution.hpp [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+                       const size_t kernelWidth,
+                       const size_t kernelHeight,
+                       const size_t strideWidth = 1,
+                       const size_t strideHeight = 1,
+                       const size_t padW = 0,
+                       const size_t padH = 0,
+                       const size_t inputWidth = 0,
+                       const size_t inputHeight = 0,
+                       const size_t numGroups = 1,
+                       const std::string& paddingType = "None");
```
Comment (shrit) [inline src/mlpack/methods/ann/layer/separable_convolution.hpp:86]: Please fix the indentation

File: src/mlpack/methods/ann/layer/separable_convolution.hpp [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+   * @param paddingType The type of padding (Valid or Same). Defaults to None.
+   */
+  SeparableConvolutionType(const size_t inSize,
+                       const size_t outSize,
+                       const size_t kernelWidth,
+                       const size_t kernelHeight,
+                       const size_t strideWidth,
+                       const size_t strideHeight,
+                       const std::tuple<size_t, size_t> padW,
+                       const std::tuple<size_t, size_t> padH,
```
Comment (shrit) [inline src/mlpack/methods/ann/layer/separable_convolution.hpp:118]: Here as well

File: src/mlpack/methods/ann/layer/separable_convolution.hpp [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+                       const size_t inputHeight = 0,
+                       const size_t numGroups = 1,
+                       const std::string& paddingType = "None");
+
+  /*
+   * Set the weight and bias term.
+   */
+  void Reset();
+  //! Clone the SeparableConvolutionType object. This handles polymorphism
+  //! correctly.
```
Comment (shrit) [inline src/mlpack/methods/ann/layer/separable_convolution.hpp:129]: Please add an empty line after the function declaration

File: src/mlpack/methods/ann/layer/separable_convolution_impl.hpp [hunk 4]
```diff
@@ -0,0 +1,478 @@
...
+  ar(CEREAL_NVP(numGroups));
+  ar(CEREAL_NVP(padding));
+
+  if (cereal::is_loading<Archive>())
+  {
+    weights.set_size((outSize * (inSize / numGroups) * kernelWidth *
+        kernelHeight) + outSize, 1);
+  }
+}
+template<
```
Comment (shrit) [inline src/mlpack/methods/ann/layer/separable_convolution_impl.hpp:448]: Please add a line after the `}`

File: src/mlpack/tests/ann/layer/separable_convolution.cpp [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+#include "../ann_test_tools.hpp"
+
+using namespace mlpack;
+
+/**
+ * Simple Separable Convolution layer test.
+ */
+TEST_CASE("SeparableConvolutionShapeAndSizeTest", "[ANNLayerTest]") {
+
+  size_t inSize = 3;
```
Comment (shrit) [inline src/mlpack/tests/ann/layer/separable_convolution.cpp:28]: Please fix the style in here

File: src/mlpack/tests/ann/layer/separable_convolution.cpp [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+                               kernelWidth,
+                               kernelHeight,
+                               strideWidth,
+                               strideHeight,
+                               padW,
+                               padH,
+                               inputWidth,
+                               inputHeight,
+                               numGroups);
+
```
Comment (shrit) [inline src/mlpack/tests/ann/layer/separable_convolution.cpp:51]: Please fix the indentation here as well

**Review by shrit — CHANGES_REQUESTED** [review]
Starting to look good, please fix the style issues and ping me when this is done, I will give it a full review

**Comment by SIDHARTH-7** [comment]
@shrit done with the indentations. Please let me know if anything else needs to be done after the review

**Review by shrit — CHANGES_REQUESTED** [review]
Please go over the style again, and do the last check, there are a couple of stuff I pointed out but it would be good to do a last check.

Once this is done I will review the test

File: src/mlpack/methods/ann/layer/separable_convolution.hpp [hunk 7]
```diff
@@ -0,0 +1,430 @@
...
+ * @file methods/ann/layer/separable_convolution.hpp
+ * @author Kartik Dutt
+ * @author Aakash Kaushik
+ * @author Sidharth
+ * 
+ * For more information, kindly refer to the following paper.
+ *
+ * @code
+ * @article{
+ *  author = {Laurent SIfre, Stéphane Mallat},
```
Comment (rcurtin) [inline src/mlpack/methods/ann/layer/separable_convolution.hpp:11]:
```suggestion
 *  author = {Laurent Sifre, Stéphane Mallat},
```

File: src/mlpack/methods/ann/layer/separable_convolution.hpp [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+#include <mlpack/prereqs.hpp>
+#include <mlpack/methods/ann/convolution_rules/naive_convolution.hpp>
+#include <mlpack/methods/ann/layer/padding.hpp>
+#include <mlpack/methods/ann/layer/layer.hpp>
+
+namespace mlpack {
+
+/**
+ * Implementation of the Separable Convolution class.The Separable Convolution class 
+ * represents a single layer of a neural network.
```
Comment (rcurtin) [inline src/mlpack/methods/ann/layer/separable_convolution.hpp:36]: This comment doesn't really explain much; do you think you can clarify it?

File: src/mlpack/methods/ann/layer/separable_convolution.hpp [hunk 9]
```diff
@@ -0,0 +1,430 @@
...
+    NaiveConvolution<ValidConvolution>,
+    arma::mat> 
+    SeparableConvolution;
+
+} // namespace mlpack
+
+// Include implementation.
+#include "separable_convolution_impl.hpp"
+
+#endif
```
Comment (rcurtin) [inline src/mlpack/methods/ann/layer/separable_convolution.hpp:429]: Can you make sure files end with newlines?  Thanks.

File: src/mlpack/tests/ann/layer/separable_convolution.cpp [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+#include "../../test_catch_tools.hpp"
+#include "../../catch.hpp"
+#include "../../serialization.hpp"
+#include "../ann_test_tools.hpp"
+
+using namespace mlpack;
+
+/**
+ * Simple Separable Convolution layer test.
```
Comment (rcurtin) [inline src/mlpack/tests/ann/layer/separable_convolution.cpp:24]: I think you haven't finished with the tests yet, but we should definitely have some more tests to ensure that the forward, backward, and gradient passes are correct.  It's likely you can adapt the existing convolution tests accordingly.  Please don't just run code from another a library and copy-paste the results; it's better to design or adapt an actual test.  If we hardcode values from another library, this is brittle: it's always possible their implementation is incorrect, or, it's also possible that they made some code choices that are not the only possible way to solve the problem.  (Think, e.g., small numerical precision errors and things of this nature.)

**Review by rcurtin — COMMENTED** [review]
Hey there @SIDHARTH-7, thanks for working on this.  I did not review closely (at all) but I skimmed through and had some simple comments.  I hope they're helpful. :+1:

File: src/mlpack/tests/ann/layer/separable_convolution.cpp (hunk 10 above)
Comment (SIDHARTH-7) [reply src/mlpack/tests/ann/layer/separable_convolution.cpp:24]: sure i will note that and add some more tests to the file 👍

Comment (SIDHARTH-7) [reply src/mlpack/tests/ann/layer/separable_convolution.cpp:24]: I have added jacobian test and numerical gradient test, If you want to add any more tests do let me know

File: src/mlpack/methods/ann/layer/separable_convolution.hpp (hunk 8 above)
Comment (SIDHARTH-7) [reply src/mlpack/methods/ann/layer/separable_convolution.hpp]: please check if the new comment is better and should we add something else too

**Comment by shrit** [comment]
@SIDHARTH-7  could you start testing this PR against the model repository and see if the Mobilenet model compiles?

**Comment by SIDHARTH-7** [comment]
Sure 👍🏼

File: src/mlpack/methods/ann/layer/separable_convolution.hpp (hunk 8 above)
Comment (rcurtin) [reply src/mlpack/methods/ann/layer/separable_convolution.hpp]: Thanks, that is indeed much improved!

**Comment by shrit** [comment]
Also, could you resolve the conflicted files?

**Comment by shrit** [comment]
I think there is still the problem with the conflicting file

**Comment by SIDHARTH-7** [comment]
> I think there is still the problem with the conflicting file

Yeah If I resolve the conflicts it is giving an error as it could not find the files 
'''
#include <mlpack/methods/ann/layer/repeat.hpp>
'''

**Comment by SIDHARTH-7** [comment]
Yeah if I resolve the conflicts it's giving an error as it could not find the files 

```
#include <mlpack/methods/ann/layer/repeat.hpp>
```
so I changed it again

**Comment by shrit** [comment]
what do you mean? they are in the master branch of mlpack, I have merged them like two weeks ago

**Comment by SIDHARTH-7** [comment]
Oh my bad I did not sync the fork sorry!

**Comment by SIDHARTH-7** [comment]
I guess the last check failed because it is not required it passes every time 
image

**Comment by shrit** [comment]
@SIDHARTH-7 hey did you test this against mobilenet model in the model repository?

**Comment by SIDHARTH-7** [comment]
@shrit The implementation of mobilenet in the models repo is old so I am refactoring it, I have finished most of the work just need to refactor it a little bit more and then I will test it locally as well as create a new PR for merging the entire thing together.

**Pull request closed without merging**